kernel/comps/virtio/src/device/input/device.rs

```diff
@@ -181,44 +185,133 @@ impl InputDevice {
         size
     }
 
-    /// Negotiate features for the device specified bits 0~23
-    pub(crate) fn negotiate_features(features: u64) -> u64 {
-        assert_eq!(features, 0);
-        0
-    }
-}
-
-impl aster_input::InputDevice for InputDevice {
-    fn handle_irq(&self) -> Option<()> {
-        // one interrupt may contains serval input, so it should loop
-        loop {
-            let Some(event) = self.pop_pending_event() else {
-                return Some(());
+    fn handle_irq(&self) {
+        // Returns ture if there may be more events to handle
+        let handle_event = |event: &EventBuf| -> bool {
+            let event: VirtioInputEvent = {
+                let mut reader = event.reader();
+                reader.read_val()
             };
+
             match event.event_type {
-                0 => return Some(()),
+                0 => return false,
                 // Keyboard
                 1 => {}
                 // TODO: Support mouse device.
-                _ => continue,
+                _ => return true,
             }
+
             let status = match event.value {
                 1 => KeyStatus::Pressed,
                 0 => KeyStatus::Released,
-                _ => return Some(()),
+                _ => return false,
             };
+
             let event = InputEvent::KeyBoard(Key::try_from(event.code).unwrap(), status);
             info!("Input Event:{:?}", event);
 
             let callbacks = self.callbacks.lock();
             for callback in callbacks.iter() {
-                callback.call((event,));
+                callback(event);
             }
+
+            true
+        };
+
+        // one interrupt may contains serval input, so it should loop
+        while self.pop_pending_event(&handle_event) {}
```

Comment:
For optimal efficiency, it is best to avoid locking `callbacks` repeatedly in this function and `event_queue` in the `pop_pending_event` function. Consider refactoring both functions to the following form.

```rust
impl InputDevice {
    // Change note: Process events in batch, instead of one by one.
    fn pop_pending_events(&self, handle_event: &impl Fn(&EventBuf) -> bool) {
        let mut event_queue = self.event_queue.lock_irq_disabled();
        while let Ok((token, _)) = event_queue.pop_used() {
            // handle the event
        }
    }

    fn handle_irq(&self) {
        //  Change note: Lock the callback per IRQ, not per event.
        let callbacks = self.callbacks.read();

        let handle_event = |event: &EventBuf| -> bool {
            // ..

            for callback in callbacks.iter() {
                callback(event);
            }
        };

        // Change note: No loop here
        self.pop_pending_events(&handle_event);
    }
}
```